### PR TITLE
Types

### PR DESCRIPTION
**Pull request by jwerle**
This PR has a bit of cleanup. It removes the `makeAPI` magic. It introduces a type system which allows for checking the type of a `Command` instance sanely. This PR also removes some useless stuff like `GeometryBuffer`. Some abstract classes were moved into `core/`

### CODE REVIEW

File: src/math/quaternion.js [hunk 1]
```diff
@@ -39,6 +40,7 @@ export class Quaternion extends Vector {
 
   constructor(x = 0, y = 0, z = 0, w = 1) {
     super(coalesce(x, 0), coalesce(y, 0), coalesce(z, 0), coalesce(w, 1))
+    assignTypeName(this, 'quaternion') // Quaternion gets special treatment
```
Comment (vipyne) [inline src/math/quaternion.js]: maybe be more specific on why we are assigning a typename

File: src/core/matrix.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * Module dependencies.
+ */
+
+import { assignTypeName } from './types'
+import { Vector } from './vector'
+
+export class Matrix extends Vector {
+  static fillna(mat, max) {
```
Comment (vipyne) [inline src/core/matrix.js]: just curious, why is it called `fillna`?

File: src/math/euler.js [hunk 3]
```diff
@@ -41,6 +42,7 @@ export class Euler extends Vector {
 
   constructor(x, y, z, order = 'xyz') {
     super(coalesce(x, 0), coalesce(y, 0), coalesce(z, 0))
+    assignTypeName(this, 'euler') // Euler gets special treatment
```
Comment (vipyne) [inline src/math/euler.js]: same comment re: comment

File: src/core/types.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+ * Assigns a type name to a given instance of anything.
+ *
+ * @public
+ * @param {Mixed} instance
+ * @param {String} typeName
+ */
+
+export const assignTypeName = (instance, typeName) => {
+  if (null == instance) { return instance }
+  if (null == typeName) { typeName == typeof instance }
```
Comment (vipyne) [inline src/core/types.js]: should this be `typeName = typeof instance` ?

File: src/core/types.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  if (false == Array.isArray(instance[$types])) {
+    instance[$types] = []
+  }
```
Comment (vipyne) [inline src/core/types.js]: can we throw in a `typeName = typeName.toLowerCase()`

File: src/core/types.js (hunk 4 above)
Comment (jwerle) [inline src/core/types.js]: oops, good catch

File: src/core/types.js (hunk 5 above)
Comment (jwerle) [inline src/core/types.js]: for sure

File: src/math/quaternion.js (hunk 1 above)
Comment (jwerle) [inline src/math/quaternion.js]: absolute

File: src/math/euler.js (hunk 3 above)
Comment (jwerle) [inline src/math/euler.js]: :+1:

**Comment by jwerle** [comment]
@vipyne yeah maybe in a subsequent PR we can modify the `incrementStat` calls

**Pull request merged**